WIP: Initial pass of Asynchronous Compilation

mgaudet opened the pull request:
This is an initial version of async compilation for Ruby+OMR. 

This code has booted Discourse, however 8a66c9e was required to pass `make test`. For more details see [this Feature proposal on bugs.ruby-lang.org](https://bugs.ruby-lang.org/issues/13512).

Commit:
Initial Async Compilation

This version of asynchronous compilation is sufficiently robust to boot
Discourse.

Signed-off-by: Matthew Gaudet <magaudet@ca.ibm.com>

Commit:
Update OMR submodule

Needed to use std::allocator.

Signed-off-by: Matthew Gaudet <magaudet@ca.ibm.com>

Commit:
Rename pyfe call to rbfe

To be accurate about what kind of call it is.

Signed-off-by: Matthew Gaudet <magaudet@ca.ibm.com>

Commit:
Comment out test cases broken by compilation thread.

Signed-off-by: Matthew Gaudet <magaudet@ca.ibm.com>

Leonardo2718 commented on rbjitglue/ruby/control/CompilationQueue.hpp:
```diff
@@ -0,0 +1,120 @@
+    * If this returns false, it is undefined behaviour
+    * to derference the parameter
+    */
+   bool pop(T& ret) { 
```
An alternative solution here is to return something like a `boost::optional` (or `std::optional` once C++17 is standard). A very simple version could also be implemented as an inner-class.

Leonardo2718 commented on rbjitglue/ruby/control/CompilationQueue.hpp:
```diff
@@ -0,0 +1,120 @@
+    * Remove every element from the queue that matches the provided predicate. 
+    *
+    * The predicate function returns true if the element is to be removed. The 
+    * predicate needs to be compatible with UnaryPredicate from <algorithms>
```
Just for reference, maybe it's better to say something like "must meet the requirements of the [Predicate concept](http://en.cppreference.com/w/cpp/concept/Predicate)"

Leonardo2718 commented on rbjitglue/ruby/control/CompilationQueue.hpp:
```diff
@@ -0,0 +1,120 @@
+    * Used to filter out compilations no longer required. 
+    */
+   template <typename CompilationPredicate> 
+   void predicateFilter(CompilationPredicate p) {
```
Minor nitpick: calling this function `filter_if` would make it more STL-like.

mgaudet replied:
I feel so burned by `remove_if` :P You're probably right, though it's not really the Testarossa style.

Leonardo2718 commented on rbjitglue/ruby/control/CompilationRegistry.hpp:
```diff
@@ -0,0 +1,98 @@
+    * the predicate using the passed mutator. 
+    */
+   template <typename CompilationPredicate, typename CompilationMutator>
+   void mutateSelectedCompilations(CompilationPredicate pred, CompilationMutator m)
```
Nitpicking again: calling this function `mutate_if` (or `transform_if` to match `std::transform`?) would make it more STL-like.

Also, stating some requirements/concepts for `CompilationPredicate`and `CompilationMutator` would be good.

Leonardo2718 commented on rbjitglue/ruby/env/GuardedCall.hpp:
```diff
@@ -0,0 +1,70 @@
+ */
+template <typename FunctionType, typename ArgumentType>
+auto GVLGuardedCall(FunctionType function, ArgumentType argument) -> decltype(function(argument))
+   {
```
Some duplication can be reduced here by using something like:

```
using ReturnType = decltype(function(argument));
```

mgaudet replied:
Totally forgot about that.

Leonardo2718 commented on rbjitglue/ruby/env/GuardedCall.hpp:
```diff
@@ -0,0 +1,70 @@
+       */
+      auto fn = [function, argument]() -> decltype(function(argument)) { 
+         return function(argument); 
+      };
```
This _looks_ to me like it could just be replaced with a call to [std::bind](http://en.cppreference.com/w/cpp/utility/functional/bind).

mgaudet replied:
I think you're right. Didn't know it existed!

mgaudet commented:
Looks like Travis has flagged some issues that didn't show up in my local testing.

I'll tackle some of @Leonardo2718's comments and see if I can reproduce some of the others locally.

Commit:
Lifetime fix for a string.

Signed-off-by: Matthew Gaudet <magaudet@ca.ibm.com>

Commit:
Simplify globals initialization

Also add ruby_vm_class_serial and ruby_vm_global_method_state to the
globals struct.

Signed-off-by: Matthew Gaudet <magaudet@ca.ibm.com>

Commit:
Disable Async Compilation by default

Will be re-enabled once the last of the bugs are worked out.

Signed-off-by: Matthew Gaudet <magaudet@ca.ibm.com>

mgaudet commented:
Current thinking is that I'm going to merge this as is (async compilation disabled by default): There is an issue with async compilation on OS/X however. I haven't quite figured out what's going on. 

I'll open an issue one this gets merged, but here are the notes I have: 

Async Compilation Failure on OS/X
------------------------------------

Can be reproduced inside LLDB by running like this repeatedly. Does not reproduce 100% of the time. 

`ENABLE_ASYNC_COMPILATION=1 OMR_JIT_OPTIONS=-Xjit:count=0,verbose,vlog=0vlog ASYNC_COMPILATION_TRACE=1 lldb -- ./ruby --disable-gems sample/test.rb`

Reproduction rate on the above command line is less than 1/10. 

Example failures: 

```
sample/test.rb:exception Process 5506 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x1014114b0)
    frame #0: 0x00000001014114b0
->  0x1014114b0: jle    0x1014114c4
    0x1014114b3: addb   (%rcx), %al
    0x1014114b5: addb   %al, (%rax)
    0x1014114b7: addb   %al, 0x14(%rax)
```

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00007fffe3b5da18 libsystem_platform.dylib`_longjmp + 40
libsystem_platform.dylib`_longjmp:
->  0x7fffe3b5da18 <+40>: ldmxcsr 0x48(%rdi)
    0x7fffe3b5da1c <+44>: cld    
    0x7fffe3b5da1d <+45>: jmpq   *0x38(%rdi)

libsystem_platform.dylib`_platform_memset_pattern4:
    0x7fffe3b5da20 <+0>:  movabsq $0x7fffffe00040, %rax     ; imm = 0x7FFFFFE00040 

(lldb) bt all
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x00007fffe3b5da18 libsystem_platform.dylib`_longjmp + 40
```

Backtraces are corrupted. 

Verbose log isn't leaving much in the way of clues... 

```
224  compiling /Users/mgaudet/tmp/rubyomr/ruby/basictest/test.rb:8:initialize       
225 #INFO:  QUEUE: Added  /Users/mgaudet/tmp/rubyomr/ruby/basictest/test.rb:422:block in <top (required)>:0x1011e5200 for compilation (queue size after 174)
226 #INFO:  QUEUE: Added  /Users/mgaudet/tmp/rubyomr/ruby/basictest/test.rb:423:block in <top (required)>:0x1011e4fc0 for compilation (queue size after 175)
227 #INFO:  QUEUE: Added  /Users/mgaudet/tmp/rubyomr/ruby/basictest/test.rb:424:block in <top (required)>:0x1011e4d80 for compilation (queue size after 176)
228 #INFO:  QUEUE: Added  /Users/mgaudet/tmp/rubyomr/ruby/basictest/test.rb:435:block in <top (required)>:0x1011e45c0 for compil
```

mgaudet commented:
Here's an update on where this is: 

1. I've been able to now reproduce on Linux. It's easiest(?) by using clang-3.8 as the build compiler. I reconfigured using the following to disable optimization in the VM, and select clang as the build compilers: ` CC=clang-3.8 CXX=clang++-3.8 PLATFORM=amd64-linux64-clang optflags=-O0 SPEC=linux_x86-64 ./configure`

2. I am able to crash the compilation threads (infrequently) without even compiling any methods: 
`for i in $(seq 1 100); do LD_LIBRARY_PATH=$PWD ASYNC_COMPILATION_TRACE=1 ./ruby -Jcount=0,verbose,exclude={*} --disable-gems sample/test.rb 2>&1 || break; done;` breaks after about 20-40 iterations on my machine. The backtrace is 

```
./home/magaudet/open/ruby_bop_notify/ruby(rb_print_backtrace+0x19) [0x5631d34ec099] vm_dump.c:685
/home/magaudet/open/ruby_bop_notify/ruby(rb_vm_bugreport+0xac) [0x5631d34ec16c] vm_dump.c:994
/home/magaudet/open/ruby_bop_notify/ruby(rb_bug_context+0x1c4) [0x5631d35791f4] error.c:426
/home/magaudet/open/ruby_bop_notify/ruby(sigsegv+0x5b) [0x5631d343e8ab] signal.c:907
/lib/x86_64-linux-gnu/libpthread.so.0 [0x7f46a5a01390]
/home/magaudet/open/ruby_bop_notify/ruby(ISEQ_ORIGINAL_ISEQ+0x14) [0x5631d353c124] ./iseq.h:64
/home/magaudet/open/ruby_bop_notify/ruby(rb_iseq_original_iseq+0x15) [0x5631d353c035] compile.c:716
/home/magaudet/open/ruby_bop_notify/librbjit.so(_ZZ14GVLGuardedCallIPFPmPK14rb_iseq_structEPS1_EDTclfp_fp0_EET_T0_ENKUlvE_clEv+0x12) [0x7f46a3f7c312] /home/magaudet/open/ruby_bop_notify/rbjitglue/ruby/env/GuardedCall.hpp:60
/home/magaudet/open/ruby_bop_notify/librbjit.so(_Z7invokerIPZ14GVLGuardedCallIPFPmPK14rb_iseq_structEPS2_EDTclfp_fp0_EET_T0_EUlvE_EPvSD_+0x1d) [0x7f46a3f7c2ed] /home/magaudet/open/ruby_bop_notify/rbjitglue/ruby/env/GuardedCall.hpp:35
/home/magaudet/open/ruby_bop_notify/ruby(rb_thread_call_with_gvl+0xa2) [0x5631d3480992] thread.c:1495
/home/magaudet/open/ruby_bop_notify/librbjit.so(_Z14GVLGuardedCallIPFPmPK14rb_iseq_structEPS1_EDTclfp_fp0_EET_T0_+0x5a) [0x7f46a3f7c2aa] /home/magaudet/open/ruby_bop_notify/rbjitglue/ruby/env/GuardedCall.hpp:63
```

This indicates that there's still some race condition occurring. My hypothesis on the above is that in between the notification from the VM firing and the compilation being created, the compilation constructor is running, but the compilation hasn't been registered yet. This means that the notification gets lost.

mgaudet commented:
So, I think there are at least two races I'm seeing. However, I have my thumb on one of them. 

Inside `compileRubyISeq` in `RubyJit.cpp`: 
```
   //In async compilation, notification could occur between here
   // and the creation of the compilation object inside compileMethod, 
   // and it would be dropped. This race can cause crashes when iseqs are collected.
   RubyMethodBlock mb(iseq, name.c_str()); // First reference iseq here.
   ResolvedRubyMethod compilee(mb); // Imagine notification happens during this constructor...
                                    //  The notification misses, and compilation continues and crashes. 
   void * startPC = compileMethod(NULL, compilee, optLevel, rc); // Compilation is registered for notifications here. 
```

mgaudet commented:
I will be on vacation for a while, which means this one will have to sit for a bit. 

I'm pondering what the right answer here is to provide the appropriate level of thread safety and interaction with the GC.

mgaudet commented:
Closing this one, as I'm going to have to rethink this a bit more when I have more time.
